Pluck Field

deckar01 opened the pull request:
Cherry pick #822 to `dev` and encapsulate `Nested`'s plucking logic in a subclass named `Pluck` with a more explicit interface.

Fixes #800

Commit:
Use the correct key when plucking fields from nested data

Commit:
Decouple plucking logic from Nested into Pluck

lafrech commented on marshmallow/fields.py:
```diff
@@ -405,22 +399,22 @@ def schema(self):
                         dump_only=self._nested_normalized_option('dump_only'))
                 else:
                     schema_class = class_registry.get_class(self.nested)
                     self.__schema = schema_class(
```
Do you think the schema instantiation could be factorized to avoid repeating the schema kwargs? Or maybe just declare the kwargs dict only once?

deckar01 replied:
Ya, they are exactly the same.

lafrech commented on marshmallow/fields.py:
```diff
@@ -445,35 +439,50 @@ def _serialize(self, nested_obj, attr, obj):
-        return ret
 
-    def _deserialize(self, value, attr, data):
+    def test_collection(self, value):
```
Why not make this method private `_test_collection`?
Same with `load` -> `_load`.

lafrech replied:
OK, I just realized you're using them in `Pluck`. Maybe that's the reason you're exposing them as public API. I don't think using them in a child class here means they need to be exposed publicly (or does it?), but maybe your point is to expose them for other potential child classes to use?

deckar01 replied:
In C++ it would be considered a "protected" member, but it was really just needed to decompose `_deserialize` into reusable pieces. Prefixing it with `_` is probably fine since it doesn't need to be used outside subclasses.

lafrech commented on tests/test_schema.py:
```diff
@@ -1913,6 +1887,48 @@ class ParentSchema(Schema):
+    def blog(self, user):
+        col1 = User(name="Mick", age=123)
+        col2 = User(name="Keith", age=456)
+        blog = Blog("Monty's blog", user=user, categories=["humor", "violence"],
```
Why not write `user=User(name="Monty", age=81)` here rather than create a dedicated single-use `user` fixture?

deckar01 replied:
It was copy pasta from copying the setup fixtures from `TestNestedSchema`. Inlining it makes sense to me.

lafrech commented on tests/test_schema.py:
```diff
@@ -1913,6 +1887,48 @@ class ParentSchema(Schema):
+
+    def test_pluck(self, blog):
+        class FlatBlogSchema(Schema):
+            name = fields.String()
```
This `name` field is never used. (Same in `FlatBlogSchema` below.)

lafrech commented on tests/test_schema.py:
```diff
@@ -1971,8 +1987,8 @@ class Meta:
 
     def test_multiple_nested_self_fields(self, user):
```
Rename as `test_multiple_pluck_self_field` and move away from `Nested` tests?

(Nitpicking: I think `rels = fields.Pluck('self', 'name', many=True, attribute='relatives')` fits in one line.)

lafrech commented:
Nothing really mandatory in my comments. This looks pretty good to me.

Thanks @deckar01 for tackling this.

lafrech commented:
When this is merged, we can add a test to raise `StringNotCollectionError` if `only` or `exclude` is passed to `Nested` as string, not list of strings.

See https://github.com/marshmallow-code/marshmallow/pull/896.

sloria commented:
Looks good to me as well. @deckar01 Can you please get this up to date? Once that's done, feel free to merge.

Commit:
Don't repeat the nested schema construction code 3 times

deckar01 commented on marshmallow/fields.py:
```diff
@@ -459,37 +443,52 @@ def _serialize(self, nested_obj, attr, obj):
+
+
+class Pluck(Nested):
+    def __init__(self, nested, field_name, **kwargs):
```
This needs docs.

deckar01 commented:
I updated the existing docs, added docs for pluck, and added an entry to the upgrading docs. This could probably use a quick review.

The code changes were all small refactors with good test coverage, so another full code review probably isn't necessary. A quick sanity check should suffice.

lafrech commented:
Great!

LGTM.

Maybe rename `Nested.load` -> `Nested._load`.
